### PR TITLE
Add metrics name linter image

### PR DESCRIPTION
**Pull request by avlitman**
Add a metric name linter based on promlint, in order to monitor metrics naming and make sure they aligned with the Prometheus metrics naming best practices.
We build and push the linter to an image in quay.io, then the image will be consumed by kubevirt operators.

jira-ticket: https://issues.redhat.com/browse/CNV-29217

### CODE REVIEW

File: test/metrics/prom-metrics-linter/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+FROM golang:1.19.2 as build
```
Comment (nunnatsa) [inline test/metrics/prom-metrics-linter/Dockerfile:1]: The go mod is with go 1.20

either change here or there so they'll match

File: test/metrics/prom-metrics-linter/custom_linter_rules.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+func CustomLinterRules(problems []promlint.Problem, mf *dto.MetricFamily, operatorName, subOperatorName string) []promlint.Problem {
+	// Check metric prefix
+	nameParts := strings.Split(*mf.Name, "_")
+
+	if nameParts[0] != operatorName {
+		problems = append(problems, promlint.Problem{
+			Metric: *mf.Name,
+			Text:   fmt.Sprintf("name need to start with '%s_'", operatorName),
```
Comment (nunnatsa) [inline test/metrics/prom-metrics-linter/custom_linter_rules.go:37]: ==> "name **must** ... "?

**Review by nunnatsa — COMMENTED** [review]
Add some inline comments

File: .github/workflows/prom-metrics-linter.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Publish operator image and generate prom-metrics-linter
+        run: |
+          docker login -u="${{ secrets.QUAY_USER }}" -p="${{ secrets.QUAY_PASSWORD }}" quay.io
+          CONTAINER_RUNTIME=docker IMG=quay.io/kubevirt/prom-metrics-linter:v0.0.1 make promlinter-build
+          CONTAINER_RUNTIME=docker IMG=quay.io/kubevirt/prom-metrics-linter:v0.0.1 make promlinter-push
```
[the remark this answers is not in the document]
Comment (avlitman) [reply .github/workflows/prom-metrics-linter.yaml:19]: it's not for every pr just if something is pushed to prom-metrics-linter folder

File: test/metrics/prom-metrics-linter/custom_linter_rules.go (hunk 2 above)
Comment (avlitman) [reply test/metrics/prom-metrics-linter/custom_linter_rules.go:37]: I saved the promlint error msg's format.

File: test/metrics/prom-metrics-linter/custom_linter_rules.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		})
+	} else if operatorName != subOperatorName && nameParts[1] != subOperatorName {
+		problems = append(problems, promlint.Problem{
+			Metric: *mf.Name,
+			Text:   fmt.Sprintf("name need to start with \"%s_%s_\"", operatorName, subOperatorName),
```
[the remark this answers is not in the document]
Comment (avlitman) [reply test/metrics/prom-metrics-linter/custom_linter_rules.go:42]: I saved the promlint error msg's format.

File: .github/workflows/prom-metrics-linter.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (avlitman) [reply .github/workflows/prom-metrics-linter.yaml:19]: we agreed to add this in a **mandatory** follow-up pr.

File: test/metrics/prom-metrics-linter/go.mod [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module github.com/kubevirt/monitoring/test/metrics/prom-metrics-linter
+
+go 1.20
```
[the remark this answers is not in the document]
Comment (avlitman) [reply test/metrics/prom-metrics-linter/go.mod:3]: Done

File: test/metrics/prom-metrics-linter/metric_name_linter.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+	// Parse metric families from the JSON representation
+	families := parseMetricFamilies(metricFamilies)
+
+	// Call customRules function to apply the custom rules to the linter
+	linter := promlint.NewWithMetricFamilies(families)
+
+	problems, err := linter.Lint()
+	if err != nil {
+		fmt.Println(err)
```
[the remark this answers is not in the document]
Comment (avlitman) [reply test/metrics/prom-metrics-linter/metric_name_linter.go:61]: Done for all

**Review by nunnatsa — APPROVED** [review]

**Pull request merged**